### PR TITLE
Hide on autostart

### PR DESCRIPTION
**Pull request by Razzeee**
This should hopefully address #228
Untested so far, will grab the artifacts from ci, as soon as they are up and test those.

Let's hope https://github.com/Teamwork/node-auto-launch/issues/31 is solved.

### CODE REVIEW

**Comment by Razzeee** [comment]
Doesn't seem to work for me on windows 10.
Can other platforms please try?
https://circleci.com/gh/mattermost/desktop/539#artifacts

**Comment by Razzeee** [comment]
Also doesn't work on win7

**Comment by jasonblais** [comment]
Any help from the community here? 

@Razzeee [Teamwork/node-auto-launch#31](https://github.com/Teamwork/node-auto-launch/issues/31) seems to reference `isHidden` function. Would that be of any use?

**Comment by Razzeee** [comment]
That's what we're trying to use. I'm just not sure if our exe understands --hidden or if they have a bug not even setting that

**Comment by jasonblais** [comment]
@yuya-oc Would you have any thoughts?

**Comment by yuya-oc** [comment]
Not applicable for OS X. In my thoughts, apps should handle `--hidden` flag because `electron` doesn't understand it, too.

**Comment by Razzeee** [comment]
Well the mac implementation isn't trying to use --hidden
https://github.com/Teamwork/node-auto-launch/blob/master/src/AutoLaunchMac.coffee

File: src/main.js [hunk 1]
```diff
@@ -19,8 +19,10 @@ if (process.platform === 'win32') {
   var cmd = process.argv[1];
   if (cmd === '--squirrel-uninstall') {
     var AutoLaunch = require('auto-launch');
+    //BUILD
     var appLauncher = new AutoLaunch({
-      name: 'Mattermost'
+      name: 'Mattermost',
+      isHidden: true
```
Comment (yuya-oc) [inline src/main.js]: I misunderstood the changelog. This part is called when uninstalling the application. So you should use `isHidden` in `settings.jsx` which enables auto-start. Then, possibly `main.js` should handle the `--hidden` flag.

**Comment by Razzeee** [comment]
Changed and now it's writing the registry key correctly. But it's like you said, electron doesn't understand --hidden for windows

**Comment by jasonblais** [comment]
Any  thoughts from the community here?

**Comment by Razzeee** [comment]
Would be nice to know if this implementation works on osx and Linux. Then we just need to figure out why windows doesn't.

**Comment by yuya-oc** [comment]
Not need to test on OS X because auto-start is not implemented. Please see #164.

**Comment by jasonblais** [comment]
I see, thanks @yuya-oc , 

@timroes @jnugh Would love to have your help checking if the desktop app stays minimized after autostart when logging in to a Linux machine.

If either of you would like to try, you can download an artifact here: https://circleci.com/gh/mattermost/desktop/539#artifacts

**Comment by timroes** [comment]
Doesn't seem to work on my system (Arch x64, pretty custom window manager - so perhaps someone with a regular Desktop Environment should test).

But looking at the generated desktop file:

```
[Desktop Entry]
Type=Application
Vestion=1.0
Name=Mattermost
Comment=Mattermost startup script
Exec=/home/timroes/apps/mattermost-desktop-linux-x64/Mattermost
StartupNotify=false
Terminal=false
```

and looking at the [source code](https://github.com/Teamwork/node-auto-launch/blob/master/src/AutoLaunchLinux.coffee#L16) of `auto-launch` when setting `isHidden` there should actually be the `--hidden` flag in the end of the Exec command, which isn't in my case.

So there might be some problems with the generation of this desktop file. Also I tried starting the above mattermost with `--hidden` directly which doesn't cause it to be hidden on my system either.

Could you point me to the source parts, where `--hidden` is read in mattermost and done somethign appropriate?

**Comment by Razzeee** [comment]
I thought the hidden would be handled by electron packager as that's responsible for building the executeables etc.  But it seems like it doesn't or is not responsible for the handling.

**Comment by jnugh** [comment]
Does not work for me (Unity).

auto-launch generates a desktop file in `~/.config/autostart` which should contains the `--hidden` param. When the default desktop file, which is being generated by `electron-(builder/packager?)`, contains the hidden flag, the app would start minimized even if it wasn't auto started.

**Comment by timroes** [comment]
Yeah. It should :-)

But apparently for me the hidden isn't in the generated desktop file. If the electron wrapper doesn't handle the hidden flag, and you don't do it, that might some the mystery why it isn't working.

**Comment by jnugh** [comment]
https://github.com/Teamwork/node-auto-launch/blob/master/src/index.coffee#L14
`isHidden` is actually mapped to `isHiddenOnLaunch`.

When running this directly from your branch, without using the artifacts, things work like expected. 3.0 is out now. maybe it's been fixed already?

**Comment by Razzeee** [comment]
Version 3.0 got rid of the ES6-Promise dependency, not sure how to fix that for us and get it building again. Changelog doesn't seem to involve a fix for our problem.

**Comment by jnugh** [comment]
https://github.com/Teamwork/node-auto-launch/issues/37 Seems like the current npm package has a bug.

**Comment by yuya-oc** [comment]
Not working in Windows 10. `--hidden` flag is passed to the app, so you have to handle the flag in `main.js`. `mainWindow` and other windows should be managed by js, so Electron should not handle this automatically. The application code should decide the behavior.

Then, please take care probably `yargs` doesn't parse all flags properly in packaged apps because the argc is different in the cases:

- electron-prebuilt: `$ electron APP_PATH args...`
- packaged-app: `$ APP args...`

**Comment by Razzeee** [comment]
Latest code should fix all but the windows installer. That will need https://github.com/Teamwork/node-auto-launch/pull/39 merged and bumped.

**Comment by yuya-oc** [comment]
Tested on Windows 10. As you said, the zip version is working properly, but the installer version is not working yet.

**Comment by jasonblais** [comment]
Adding a `pending` label until https://github.com/Teamwork/node-auto-launch/pull/39 is merged

**Comment by Razzeee** [comment]
@timroes @jnugh 
Would be nice to know if it already works for linux like this: https://circleci.com/gh/mattermost/desktop/649#artifacts

**Comment by jnugh** [comment]
Minimized autostart works for me :)

**Comment by timroes** [comment]
@Razzeee works for me too (custom window manager, Archlinux x64), but brings me back to another problem, which I will open another ticket for (skip taskbar hints).

**Comment by Razzeee** [comment]
Upstream just merged and release https://github.com/Teamwork/node-auto-launch/pull/39
I've updated this PR to include it, so maybe we can still make it into 1.4 :rocket: 

Let's wait for the artifacts

**Comment by jasonblais** [comment]
Awesome! I tried looking for the artifacts earlier today [in CircleCI](https://circleci.com/gh/mattermost/desktop) but couldn't find them.. 

@yuya-oc are they located elsewhere?

Added for v1.4 so we can speed up the testing.

**Comment by Razzeee** [comment]
https://circleci.com/gh/Razzeee/desktop/17#artifacts

**Comment by jasonblais** [comment]
Thanks!

Just tested (Windows 10, 64-bit) and it didn't auto-start for me when I restarted my machine.

I have `Start app on login.` checked.

**Comment by yuya-oc** [comment]
I have tested .zip and setup.exe on Windows 10 64-bit, and .deb on Ubuntu 14.04 64-bit. They are working well.

@jasonblais Would you check again?

**Comment by jasonblais** [comment]
Hmm okay, maybe the issue was on my end.

**Comment by jasonblais** [comment]
Tested; works on Windows 10 after disabling and re-enabling the `start app on login` setting.

@yuya-oc helped troubleshoot and likely related to me running a v1.3 version where auto-start might have been disabled. 

The setting is not overwritten unless you disable and re-enable it in the settings UI (if you simply click 'Save' on the settings page without a change, the option will remain to what it was before in the registry). Investigating whether an easy fix for v1.4/3.4 or if we leave it for the next version.

For this particular PR however, we're good to go. Thanks everyone!

**Comment by yuya-oc** [comment]
So I will change the behavior when saving options in settings. It would always update the registry info. Thanks!

**Comment by Razzeee** [comment]
Please do that in a pr and give me some heads up, as I have some problems
with other apps that work like that. Due to macaffee doing stupid stuff.

**Comment by yuya-oc** [comment]
@Razzeee I created new PR #283. Would you write about the problem you met in detail?

**Comment by jasonblais** [comment]
Thanks @Razzeee, agree that this will need some careful testing.  Let's take it in the next release.

Thanks @yuya-oc for the pull request! We'll test it out

**Pull request merged**